backend/src/services/auto_trader.py

```
@@ -33,238 +23,431 @@ def __init__(self, socketio=None, data_fetcher_service=None, ai_analysis_service
+                current_balance = await self.wallet_service.get_token_balance(token_address)
+                if current_balance <= 0:
+                    tokens_to_remove.append(token_address)
                     continue
 
+                # Sync balance if it changed
+                if abs(current_balance - details.get('amount_tokens', 0)) > 0.000001:
+                    details['amount_tokens'] = current_balance
+                    await asyncio.to_thread(save_position, details)
+
+                current_token_data = await self.data_fetcher_service.get_token_by_address(token_address)
+                if not current_token_data: continue
```

Ruff (0.15.12)

[error] Multiple statements on one line (colon)

(E701)